Schema change: add label->tag mapping table.

cben opened the pull request:
Extracted schema+model from ongoing work on #7605.
cc @Fryguy @kbrock @zeari @moolitayer @simon3z 
We'd like to get this, or similar, merged before the freeze.

Summary:

- The goal is assigining Tags for some but not all labels.
  [Reminder: labels are stored as CustomAttribute with polymorphic resource, string name & value; unlike Tagging, there is no central object like Tag that all labels with same name & value point to.]

- I'm matching on resource_type (null means any), name (required), value (null means any).
  - Any-value mapping rows don't map to a single tag, but one tag per value (created on demand).
    In such mappings `tag_id` means the category under which to create tags.

[I UPDATED this description to reflect current status.]

cben commented:
@miq-bot label enhancement providers/containers

Oh, the original issue with high-level goals is #7460.

simon3z commented:
@cben please from `CustomAttributeToTagMapping` remove all the code and leave only the relationship definitions e.g. `belongs_to` etc.

Also drop from this PR the changes of `app/models/classification.rb` (leave them in the other PR).

Thanks.

cben commented:
Done.

cben commented:
Oh, and naming/bikeshedding question:
I called it "custom_attribute_to_tag_mapping" but in the forseeable future we're only interested in section="labels", source="kubernetes" attributes, and I'm not storing those.  Is that good?  Should I name it "label_to_tag_mapping"?  Should I add section & source columns?

simon3z commented:
> Oh, and naming/bikeshedding question:
> I called it "custom_attribute_to_tag_mapping" but in the forseeable future we're only interested in section="labels", source="kubernetes" attributes, and I'm not storing those. Is that good? Should I name it "label_to_tag_mapping"? Should I add section & source columns?

@cben thanks for splitting. Yes, the naming should be updated, since this is specific for containers, it should be something like `ContainerLabelTagMappings`, or even `ContainerLabelMappings`. cc @blomquisg @Fryguy for naming suggestions.

simon3z commented on db/migrate/20160412085547_create_custom_attribute_to_tag_mappings.rb:
```diff
@@ -0,0 +1,13 @@
+  def change
+    create_table :custom_attribute_to_tag_mappings do |t|
+      t.string :resource_type
+      t.string :name
```
@cben is this the label name? Maybe `label_name` is clearer.

simon3z commented on db/migrate/20160412085547_create_custom_attribute_to_tag_mappings.rb:
```diff
@@ -0,0 +1,13 @@
+    create_table :custom_attribute_to_tag_mappings do |t|
+      t.string :resource_type
+      t.string :name
+      t.text :value
```
@cben probably `label_value`

cben commented:
Renamed the table, and renamed the columns (including type).

I tried to use `label_key` in place of `label_name` (because we know the `name` of a CustomAttribute that represents a label is its "key" in kubernetes terminology), but was unable to even write a coherent comment explaining what I'm matching on, so went with `label_name`.

cben commented:
Oh, a question I completely forgot: does this table need any indexes?
To be honest, I don't understand DBs enough to answer it myself.
Might depend on question 2 in top message: may I hold the whole table in memory, or do I query for every entity with some targeted fancy join?

simon3z commented on db/migrate/20160414185708_create_container_label_tag_mappings.rb:
```diff
@@ -0,0 +1,13 @@
+class CreateContainerLabelTagMappings < ActiveRecord::Migration[5.0]
+  def change
+    create_table :container_label_tag_mappings do |t|
+      t.string :labeled_resource_type
```
@Fryguy not sure if this follows the rails/miq standards.

cben replied:
Just to clarify, `labeled_resource_type` is not trying to a polymorphic association, 
it's just a string (or NULL) constraining the value of `resource_type` on the label.
[Motivation ex: one might want to map `zone` labels to tags only on container nodes and not all entities.  Though I expect it will be mostly unconstrained.]

moolitayer replied:
Does it make sense for labeled_resource_type to be named resource_type?

cben replied:
I called it `labeled_` for consistency with @simon3z's suggestion of `label_name` and `label_value`; I'm happy to adopt whatever naming people tell me :-)

blomquisg replied:
@cben, can you give a quick example of a `labeled_resource_type`?  My eyes are still tricking me into thinking this has *something* to do with polymorphic relationships, even though I know it doesn't.

cben replied:
In theory same label on say nodes vs pods might have different meaning, and deserve different treatment.
Let's say I want to tag only container nodes with label region=infra but not other entity types that have that label, so I'll set `labeled_resource_type: "ContainerNode", label_name: "region", label_value: "infra"`.

Admittedly that's a silly example.
In practice, I don't have an example of "region" having a different meaning on other entities,
or of *any* case where constraining by resource_type is important!

[See https://gist.github.com/cben/2619583f777ddb480a10#file-labels-txt for actually occuring labels I've seen, and https://github.com/ManageIQ/manageiq/pull/7605#issuecomment-206920881 for my analysis which labels a user might want to map to tags.]

But since kubernetes/openshift users may invent whatever labels they see fit,
it seemed like something the person defining a mapping _could_ want.

@simon3z Please chime in if you have better idea if per-type mapping is useful.

cben replied:
(The reason it sounds like a polymorphic relationship is that on CustomAttribute, resource_{id,type} _is_ a polymorphic relationship.  Better naming ideas welcome.)

Fryguy replied:
I try to avoid `_type` column names to avoid confusion, but there's nothing wrong with it.  I usually take a hard stance against `_id` named columns though.

Fryguy replied:
Actually, in this case, since it related to an actual class, perhaps `labeled_resource_class` might be a better name (but I'm ok with _type too)

simon3z commented:
> Oh, a question I completely forgot: does this table need any indexes?

I'd say that we don't need them at this point. cc @Fryguy 

> Might depend on question 2 in top message: may I hold the whole table in memory, or do I query for every entity with some targeted fancy join?

It seems that this question about the 2 options is not strictly related to this PR (would it affect the code here?). We can continue the discussion on the other PR which I believe is the only one relevant. (But correct me if I missed something).

Overall LGTM :+1: . @cben please work with @zeari @alongoldboim @moolitayer to get this ACK'd. Probably they'll need a little bit of background on what we're doing here.

moolitayer commented on app/models/tag.rb:
```diff
@@ -4,6 +4,8 @@ class Tag < ApplicationRecord
   virtual_has_one :category,       :class_name => "Classification"
   virtual_has_one :categorization, :class_name => "Hash"
 
+  has_one :container_label_tag_mapping # Only for tags assigned by ContainerLabelTagMapping.
```
@cben Can you please explain what this relationship is for? I can't understand it from the pr

cben replied:
I just realized this is wrong, should be has_many — one may map multiple kinds of labels to one tag.

This is just the inverse relation to the mapping's `belongs_to :tag` from the mapping.  It has no effect on the actual schema - only the mapping table has a `tag_id`.

Yes, the reason it's useful is very unclear in this cut-down PR.  
When refreshing an entity, I should be able to clear auto-assigned tags that are no longer relevant (because label was changed/removed).  I can't wipe out _all_ tags, only ones I assigned.  Which are precisely tags figuring in at least one mapping tag_id.
Googling up on Rails, I discovered `Tag.joins(:custom_attribute_to_tag_mapping)` gives me them, but I must declare the association from Tag side for it to work.

cben commented:
Yeah, question 2 is irrelevant here (as long as it doesn't lead to add_index), will discuss separately.

Pushed -f again (fixed has_one to has_many in tag.rb).

alongoldboim commented on db/migrate/20160414185708_create_container_label_tag_mappings.rb:
```diff
@@ -0,0 +1,13 @@
+      t.string :label_name
+      t.text :label_value
+      t.bigint :category_tag_id
+      t.bigint :tag_id
```
Gave this comment on the other pr as well.
* I like to use
```ruby
t.belongs_to :tag, :type => :bigint
```

It has no difference but its clearer.

* I think you can drop the "category_tag_id" if the tag is connected to its category, i assume you can get it from the tag (.classifaction/.category) and place a ```:through``` ?

cben replied:
@alongoldboim commented on https://github.com/ManageIQ/manageiq/pull/7605#discussion_r60007351

> - I like to use 
>
>   ```
>   t.belongs_to :tag, :type => :bigint
>   ```
>
>   It has no difference but its clearer.

I've considered `t.references` but saw practically none of our migrations used it — but I now see the `t.belongs_to` alias is widespread.
The stopper was that based on some googling I can't use this style with 2 names referring into the same tables [http://stackoverflow.com/a/22384289/239657 is closest but requires an extra gem].
IMHO, mixing:
```
t.belongs_to :tag, :type => :bigint
t.bigint :category_tag_id
```
would be less readable?

> - I think you can drop the "category_tag_id" if the tag is connected to its category, i assume you can get it from the tag (.classifaction/.category) and place a :through ?

That's true for (name, value) -> specific tag mappings.  For these the code only looks at `tag`; `category_tag` doesn't matter and if it did it could be computed from `tag`.

However I also have open-ended (name, *) -> * mappings where the tag is created on demand.
These rows have no `label_value` and no `tag`, but need some way to specify a category under which the tag should be created — that's what `category_tag` does.
[My current implementation does create a specific-tag mapping row every time it creates a specific tag.  But it still needs the open-ended row in case it encounters new values in the future.]

- I could abuse `tag` to double-serve as the parent tag or specific tag depending on whether `label_value` is set.
  I thought it'd be messy.

P.S. Could you please take a look at the comment in app/models/container_label_tag_mapping.rb
 and tell me if they explain this well?

alongoldboim replied:
```ruby 
 belongs_to :tag, :type => :bigint
 belongs_to :tag_category , :class_name => 'Tag', :inverse_of => :tag, :type => :bigint
```

Think this should do.

cben commented:
Addressed in-person review by @alongoldboim @moolitayer:
Merged category_tag_id into tag_id, as I need one or the other but not at the same time.

simon3z commented:
LGTM :+1:

Ready for merge unless critical issues in the review cc @chessbyte @Fryguy @blomquisg

moolitayer commented:
LGTM :+1:

simon3z commented:
LGTM :+1: ready for merge cc @chessbyte @Fryguy @blomquisg

alongoldboim commented:
LGTM :+1:

cben commented:
Pushed -f rubocop change (validates_presence_of -> `validates :label_name, :presence => true`).

Fryguy commented on db/migrate/20160414185708_create_container_label_tag_mappings.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateContainerLabelTagMappings < ActiveRecord::Migration[5.0]
+  def change
+    create_table :container_label_tag_mappings do |t|
```
I'm not sure this should be specific to containers, but I think I'm ok with it as a first pass, and then if we try to implement this for say, amazon, we can always rename it.

Fryguy commented on db/migrate/20160414185708_create_container_label_tag_mappings.rb:
```diff
@@ -0,0 +1,12 @@
+      t.string :labeled_resource_type
+      t.string :label_name
+      t.text :label_value
+      t.belongs_to :tag, :type => :bigint
```
Minor, but align the symbols for readability.

cben replied:
Too late as you already merged?  Let me know if you want a followup PR for any of your comments.

Fryguy commented:
@cben I like the concept and the table as is, so I'm going to merge it.  I think the table should be renamed to be more provider generic, but I don't see us adding label-tag mapping for other providers for the darga release.

cben commented:
Thanks!
Ironically, I initially named it CustomAttributeToTagMapping but then we figured container labels is all we have a use for currently :-)

I wasn't even aware we have labels from anything other than kubernetes—being new to the project I have a very myopic view, basically only familiar with some of the containers functionality...
